be/src/exec/pipeline/exchange/exchange_parallel_merge_source_operator.cpp

```diff
@@ -0,0 +1,121 @@
+// limitations under the License.
+
+#include "exec/pipeline/exchange/exchange_parallel_merge_source_operator.h"
+
+#include "exec/sort_exec_exprs.h"
+#include "runtime/data_stream_mgr.h"
+#include "runtime/data_stream_recvr.h"
+#include "runtime/exec_env.h"
+
+namespace starrocks::pipeline {
+
+Status ExchangeParallelMergeSourceOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(SourceOperator::prepare(state));
+    auto* factory = down_cast<ExchangeParallelMergeSourceOperatorFactory*>(_factory);
+    _stream_recvr = factory->get_stream_recvr(state, _unique_metrics);
+    _merger = factory->get_merge_path_merger(state);
+    _merger->bind_profile(_driver_sequence, _unique_metrics.get());
+    return Status::OK();
+}
+
+bool ExchangeParallelMergeSourceOperator::has_output() const {
+    if (_is_finished) {
+        return false;
+    }
+    if (_merger->is_current_stage_finished(_driver_sequence)) {
+        return false;
+    }
+    if (_merger->is_pending(_driver_sequence)) {
+        return false;
+    }
+    return true;
+}
+
+bool ExchangeParallelMergeSourceOperator::is_finished() const {
+    if (_limit < 0) {
+        return _is_finished;
+    } else {
+        return _num_rows_returned >= _limit || _is_finished;
+    }
+}
+
+Status ExchangeParallelMergeSourceOperator::set_finishing(RuntimeState* state) {
+    _is_finished = true;
+    auto* factory = down_cast<ExchangeParallelMergeSourceOperatorFactory*>(_factory);
+    factory->close_stream_recvr();
+    return Status::OK();
+}
+
+StatusOr<ChunkPtr> ExchangeParallelMergeSourceOperator::pull_chunk(RuntimeState* state) {
+    // TODO process limit
+    auto chunk = _merger->try_get_next(_driver_sequence);
+
+    if (_merger->is_finished()) {
+        _is_finished = true;
+    }
+
+    eval_runtime_bloom_filters(chunk.get());
+    return std::move(chunk);
+}
+
+Status ExchangeParallelMergeSourceOperatorFactory::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(OperatorFactory::prepare(state));
+    RETURN_IF_ERROR(_sort_exec_exprs->prepare(state, _row_desc, _row_desc));
+    RETURN_IF_ERROR(_sort_exec_exprs->open(state));
+    return Status::OK();
+}
+
+OperatorPtr ExchangeParallelMergeSourceOperatorFactory::create(int32_t degree_of_parallelism, int32_t driver_sequence) {
+    ++_stream_recvr_cnt;
+    return std::make_shared<ExchangeParallelMergeSourceOperator>(this, _id, _plan_node_id, driver_sequence, _num_sender,
+                                                                 _row_desc, _sort_exec_exprs, _is_asc_order,
+                                                                 _nulls_first, _offset, _limit);
+}
+
+void ExchangeParallelMergeSourceOperatorFactory::close(RuntimeState* state) {
+    _sort_exec_exprs->close(state);
+    OperatorFactory::close(state);
+}
+
+DataStreamRecvr* ExchangeParallelMergeSourceOperatorFactory::get_stream_recvr(
+        RuntimeState* state, const std::shared_ptr<RuntimeProfile>& profile) {
```

Comment:
will this interface be accessed by multi thread during prepare phase? should we ensure it thread safety?

Reply:
prepare stage is working at single thread mode. So no concurrency issue here.